### PR TITLE
Added option to let user accurately emulate sound with speed boost

### PR DESCRIPTION
**Pull request by mtjohnson47**
The higher pitch when boosting FPS annoys me more than the crackling.  I'm sure I'm not the only person who feels the same way.
Also, the feature freeze doesn't apply to this because it's just adding a different workaround to the audio bug.

---
This change is [Reviewable](https://reviewable.io/reviews/dolphin-emu/dolphin/3835)

### CODE REVIEW

**Comment by Parlane** [comment]
@dolphin-emu-bot rebuild

File: Source/Core/Core/HW/SystemTimers.cpp [hunk 1]
```diff
@@ -102,6 +102,10 @@ static void DSPCallback(u64 userdata, s64 cyclesLate)
 static void AudioDMACallback(u64 userdata, s64 cyclesLate)
 {
 	int period = s_cpu_core_clock / (AudioInterface::GetAIDSampleRate() * 4 / 32);
+	if (SConfig::GetInstance().bEmulationSpeedFixAudio &&
```
Comment (degasus) [inline Source/Core/Core/HW/SystemTimers.cpp]: Oh, please do not touch the AI. AI is for emulation. I think you want to disable this: https://github.com/dolphin-emu/dolphin/blob/master/Source/Core/AudioCommon/Mixer.cpp#L62

File: Source/Core/DolphinWX/Config/GeneralConfigPane.cpp [hunk 2]
```diff
@@ -40,7 +40,7 @@ GeneralConfigPane::GeneralConfigPane(wxWindow* parent, wxWindowID id)
 void GeneralConfigPane::InitializeGUI()
 {
 	m_throttler_array_string.Add(_("Unlimited"));
-	for (int i = 10; i <= 200; i += 10) // from 10% to 200%
+	for (int i = 10; i <= 400; i += 10) // from 10% to 400%
```
Comment (degasus) [inline Source/Core/DolphinWX/Config/GeneralConfigPane.cpp]: unrelated

File: Source/Core/Core/HW/SystemTimers.cpp (hunk 1 above)
Comment (mtjohnson47) [inline Source/Core/Core/HW/SystemTimers.cpp]: Actually, with that commented out, and mine disabled, it sounds 100% correct when reaching full 200% FPS, but stutters before then.
With that commented out, and mine enabled, audio plays faster than normal speed, but never stutters.

Any way to edit a pull request?  And sorry, I don't know much about dolphin's internals, and where I edited is what I managed to find with Ctrl-F and Ctrl-Click

**Comment by JMC47** [comment]
Anything touching audio interface absolutely would be subject to the feature freeze.

File: Source/Core/Core/HW/SystemTimers.cpp (hunk 1 above)
Comment (degasus) [inline Source/Core/Core/HW/SystemTimers.cpp]: Yeah, you can add new commits, or ammend your new changes and force-push the updated commit.

**Comment by phire** [comment]
Yeah, such hacks should probably be confined In audiocommon (mixer.cpp), automatically adjusting the expected sample rate based on emulation speed (which already resamples it to 48khz) rather than touching AudioInterface (which is visible to the emulated game)

File: Source/Core/Core/HW/SystemTimers.cpp (hunk 1 above)
Comment (degasus) [inline Source/Core/Core/HW/SystemTimers.cpp]: But it's expected to have perfect audio with this hack here and the disabled mixer, but this will break many games in a very unepexceted way. The expected ways are: audio-video desyncs in movies and no increase in gamespeed on several games.
IMO the better way would be to apply some kind of timestretching as we do within the OpenAL backend. It would still sounds worse than your proposed hack, but it won't break lots of games.

**Comment by JosJuice** [comment]
This is definitely a feature in my opinion, and even if it wouldn't count as one, the string freeze prevents this PR from being merged. It's not appropriate for 5.0.

**Comment by Helios747** [comment]
+1 for pushing this til after 5.0

**Comment by TruckJitsu** [comment]
If you cut the samples in half in AudioInterface.cpp, you fix the sound issues from doubling the FPS/emulation. If anything, there should just be a check box to "Cut Samples in Half for 120FPS" or something. Unless I'm totally misunderstanding this...

**Comment by JosJuice** [comment]
@TruckJitsu Messing around with AudioInterface like that is likely to break many games. In Dolphin, we try to not add built-in hacks that only benefit certain games, because they make maintenance harder in the long run, can lead to users enabling them in inappropriate situations and getting confused about why games don't work correctly, along with other problems. Without having that as a general policy, I don't think Dolphin would've come as far as it has today.

You're referring to a 120 FPS hack (as opposed to just wanting to run the game at double speed), right? If you want to make a game run at a higher framerate, the way to go is making a game modification that makes it run at a higher framerate, not making a game modification that halves the game speed then adjusting Dolphin's emulation speed to counteract that then hacking Dolphin's audio sample rate to counteract that. The latter is very hacky and far from how actual consoles work, and is also harder for users to set up.

If you do just want to run the game at double speed, it's still very hacky. Refer to what degasus said earlier.

**Comment by RisingFog** [comment]
Now that 5.0 is released, what's the status of this PR?

**Comment by JosJuice** [comment]
This PR will not get merged unless it gets rewritten into something entirely different. Considering that we've had the post-5.0 reformatting, opening a new PR is probably the easiest way to go if the PR author wants to do something like what degasus suggested.
